atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -348,3 +364,125 @@ def get_signing_intent(self):
+        matches = []
+        # py2 glob.glob doesn't support recursive, hence os.walk & fnmatch
+        # py3 can use: glob.glob(os.path.join(unpack_dir, '**', exclude), recursive=True)
+        for root, dirnames, filenames in os.walk(unpack_dir):
+            for entry in dirnames + filenames:
+                full_path = os.path.join(root, entry)
+
+                for exclude in exclude_list:
+                    if full_path.endswith(exclude):
+                        matches.append(full_path)
+                        break
+        return matches
+
+    def _remove_excluded_matches(self, matches):
+        for entry in matches:
+            if os.path.exists(entry):
+                if os.path.isdir(entry):
+                    self.log.debug("Removing excluded directory %s", entry)
+                    shutil.rmtree(entry)
+                else:
+                    self.log.debug("Removing excluded file %s", entry)
+                    os.unlink(entry)
+
+    def exclude_files_from_remote_sources(self, remote_sources_map, remote_sources_dir):
+        """
+        :param remote_sources_map: dict, keys are filenames of sources from cachito,
+                                         values are url with json from cachito
+        :param remote_sources_dir: str, dir with downloaded sources from cachito
+        """
+        src_config = get_source_container(self.workflow, fallback={})
+        exclude_list_url = src_config.get('excludelist_content')
+
+        if not exclude_list_url:
+            self.log.debug('no "excludelist_content" defined, not excluding any '
+                           'files from remote sources')
+            return
+
+        request_session = get_retrying_requests_session()
+
+        exclude_list = self._get_exclude_list(request_session, exclude_list_url)
+
+        # key: full path to source archive, value: cachito json
+        full_remote_sources_map = self._create_full_remote_sources_map(request_session,
+                                                                       remote_sources_map,
+                                                                       remote_sources_dir)
+        for remote_archive, remote_json in full_remote_sources_map.items():
+            unpack_dir = remote_archive + '_unpacked'
+
+            with tarfile.open(remote_archive) as tf:
+                tf.extractall(unpack_dir)
+
+            delete_app = self._check_if_package_excluded(remote_json, exclude_list, remote_archive)
+
+            # if any package in cachito json matched excluded entry,
+            # remove 'app' from sources, except 'app/vendor' when exists
+            if delete_app and os.path.exists(os.path.join(unpack_dir, 'app')):
+                self._delete_app_directory(remote_sources_dir, unpack_dir, remote_archive)
+
+            # search for excluded matches
+            matches = self._get_excluded_matches(unpack_dir, exclude_list)
+
+            self._remove_excluded_matches(matches)
```

Comment:
My understand of the lines from 464 to 477 is they are filtering out files from extracted cachito archive file by finding out excluded items. May I suggest another way to implement, that is to use use `members` argument of `TarFile.extractall`, and `TarFile.getmembers` is helpful to get the full list of names within the archive file.

In this way, we don't have to extract the whole archive file and do all the necessary IO operations through the whole unpack directory tree. IIRC, I was told that there might be much big archive files in cachito. However, I'm not sure if that would affect source container build.

Basically, the workflow could look like:

```
def filter_func(members):
     yield the included member only

tarfile.extractall(unpack_dir, members=filter_func(tarfile))
remove remote_archive
create new remote_archive from unpack_dir
```

Reply:
well I need to first unpack archive,
delete matched dirs/files from it 
and then pack it again 

so I don't see how useful is it 

also keep in mind, that we will be deleting just few things from the archive